clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -290,18 +317,47 @@ public void poll(long timeout) throws IOException {
 
         clear();
 
-        if (hasStagedReceives() || !immediatelyConnectedKeys.isEmpty())
+        boolean dataInBuffers = !keysWithBytesBuffered.isEmpty();
+        
+        if (hasStagedReceives() || !immediatelyConnectedKeys.isEmpty() || (madeProgressLastPoll && dataInBuffers))
             timeout = 0;
 
+        if (!memoryPool.isOutOfMemory() && outOfMemory) {
+            //server has recovered from memory pressure. unmute any channel not explicitly muted for other reasons
+            log.trace("Broker no longer low on memory - unmuting incoming sockets");
+            for (KafkaChannel channel : channels.values()) {
```

Comment:
explicitlyMutedChannels are channels muted because they already have an outstanding request in progress. we never want to service them until they are (explicitely) unmuted and taken out of the set ?

so this loop iterates over _ALL_ channels, unmuting anything that _ISNT_ in explicit.